// smartacjs/lib/scNamedLog.js
/**
 * Copyright © 2015 Smartac Inc. All rights reserved.
 * Created by Simon on 2015/9/16.
 */
var sc=require('./scCommon')
// 定义：命名日志
function NamedLog(module,name,obj)
{
    this.reset(module,name,obj);
    if (!!!this._obj)
    {
        this._obj=console;
    }
}
// 重置
NamedLog.prototype.reset=function(module,name)
{
    this._module=module;
    this._name=name;
    this._prefix="";
    if (this._module){
        this._prefix+="["+this._module+"]";
    }
    if (this._name){
        this._prefix+="["+this._name+"]";
    }
}
NamedLog.prototype._format=function(args)
{
    var msg=sc.formatArgs(args);
    if (msg==="null") throw new Error("log message is:null");
    return this._prefix+msg;
}
// 输出
NamedLog.prototype.error=function(){
    var msg=this._format(Array.prototype.slice.call(arguments));
    this._obj.error(msg);
}
NamedLog.prototype.warn=function(){
    var msg=this._format(Array.prototype.slice.call(arguments));
    this._obj.warn(msg);
}
NamedLog.prototype.info=function(){
    var msg=this._format(Array.prototype.slice.call(arguments));
    this._obj.info(msg);
}
NamedLog.prototype.debug=function(){
    var msg=this._format(Array.prototype.slice.call(arguments));
    this._obj.debug(msg);
}
NamedLog.prototype.trace=function(){
    var msg=this._format(Array.prototype.slice.call(arguments));
    this._obj.log(msg);
}
NamedLog.prototype.log=NamedLog.prototype.trace;
// 导出定义
exports.create=function(module,name,obj)
{
    return new NamedLog(module,name,obj);
}
